cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/trigger/AndTrigger.java

```diff
@@ -46,4 +49,33 @@ public boolean isSatisfied(ProgramSchedule schedule, List<Notification> notifica
   public List<TriggerInfo> getTriggerInfos(TriggerInfoContext context) {
     return getUnitTriggerInfosAddRuntimeArgs(context);
   }
+
+  @Nullable
+  @Override
+  public Trigger updateTriggerWithDeletedProgram(ProgramId programId) {
+    List<SatisfiableTrigger> updatedTriggers = new ArrayList<>();
+    for (Trigger trigger : getTriggers()) {
+      if (trigger instanceof ProgramStatusTrigger &&
+        programId.equals(((ProgramStatusTrigger) trigger).getProgramId())) {
+        // this program status trigger will never be satisfied, so the current AND trigger will never be satisfied
+        return null;
+      }
+      if (trigger instanceof co.cask.cdap.internal.app.runtime.schedule.trigger.AbstractCompositeTrigger) {
+        Trigger updatedTrigger = ((co.cask.cdap.internal.app.runtime.schedule.trigger.AbstractCompositeTrigger) trigger)
+          .updateTriggerWithDeletedProgram(programId);
+        if (updatedTrigger == null) {
+          // the updated composite trigger will never be satisfied, so the AND trigger will never be satisfied
+          return null;
+        }
+        // add the updated composite trigger into updatedTriggers
+        updatedTriggers.add((SatisfiableTrigger) updatedTrigger);
```

Comment:
this is guaranteed by the creation of schedules. However, `getTriggers()` is inherited from `ProtoTrigger.AbstractComposite` which cannot return `SatisfiableTrigger`

Reply:
This inheritance hierarchy seems strange....

It's bad to make these assumptions when you don't need to, as it tightly couples all these classes. If somebody changes the constructor, this will suddenly break. You can add a method to AbstractCompositeTrigger (not the one in proto) to return the set of SatisfiableTriggers.